adhocracy4/comments_async/static/comments_async/comment_box.jsx

```diff
@@ -93,167 +87,140 @@ export default class CommentBox extends React.Component {
+              setLoading(false)
             } else {
-              this.fetchComments()
+              fetchComments()
             }
           } else {
-            if (this.state.anchoredCommentId) {
+            if (props.anchoredCommentId) {
               /* display something like: django.gettext('We are sorry, this comment does not exist.')
                * probably using a modal
                */
             }
-            this.setState(
-              {
-                comments: data.results,
-                nextComments: data.next,
-                commentCount: data.count,
-                loading: false,
-                hasCommentingPermission: data.has_commenting_permission,
-                wouldHaveCommentingPermission: data.would_have_commenting_permission,
-                projectIsPublic: data.project_is_public,
-                useTermsOfUse: data.use_org_terms_of_use,
-                agreedTermsOfUse: data.user_has_agreed
-              }
-            )
+            setComments(data.results)
+            setNextComments(data.next)
+            setCommentCount(data.count)
+            setLoading(false)
+            setHasCommentingPermission(data.has_commenting_permission)
+            setWouldHaveCommentingPermission(data.would_have_commenting_permission)
+            setProjectIsPublic(data.project_is_public)
+            setUseTermsOfUse(data.use_org_terms_of_use)
+            setAgreedTermsOfUse(data.user_has_agreed)
           }
         }
       )
-  }
-
-  // remove auto scroll
-  componentWillUnmount () {
-    window.removeEventListener('scroll', this.handleScroll)
-  }
+  }, [])
 
   // handles update of the comment state
   // called in handleCommentSubmit, handleCommentModify, handleCommentDelete,
   // handleHideReplyError, handleHideEditeError
-  updateStateComment (index, parentIndex, updatedComment) {
-    const comments = this.state.comments
+  function updateStateComment (index, parentIndex, updatedComment) {
     const diff = {}
     if (typeof parentIndex !== 'undefined') {
       diff[parentIndex] = { child_comments: {} }
       diff[parentIndex].child_comments[index] = { $merge: updatedComment }
     } else {
       diff[index] = { $merge: updatedComment }
     }
-    this.setState({
-      comments: update(comments, diff)
-    })
+    setComments(update(comments, diff))
+  }
+
+  function addComment (parentIndex, comment) {
+    let diff = {}
+    let newCommentCount = commentCount
+    if (typeof parentIndex !== 'undefined') {
+      diff[parentIndex] = {
+        child_comments: { $push: [comment] },
+        $merge: {
+          replyError: false,
+          errorMessage: undefined
+        }
+      }
+    } else {
+      diff = { $unshift: [comment] }
+      newCommentCount++
+      setMainError(undefined)
+    }
+    setComments(update(comments, diff))
+    setCommentCount(newCommentCount)
+  }
+
+  function setReplyError (parentIndex, index, message) {
```

Comment:
I've added a bunch of convenient functions for the error handling as I found it hard to read